Chrisnas/docs 2799 dotnet profiler (#12736)
DataDog/documentation: content/en/tracing/profiler/enabling/dotnet.md

kind: Documentation
code_lang: dotnet
type: multi-code-lang
code_lang_weight: 60
further_reading:
    - link: 'getting_started/profiler'
      tag: 'Documentation'
      text: 'Getting Started with Profiler'
    - link: 'tracing/profiler/search_profiles'
      tag: 'Documentation'
      text: 'Learn more about available profile types'
    - link: 'tracing/profiler/profiler_troubleshooting'
      tag: 'Documentation'
      text: 'Fix problems you encounter while using the profiler'
---

{{< site-region region="us5" >}}
<div class="alert alert-warning">
  The Continuous Profiler is not available for the Datadog {{< region-param key="dd_site_name" >}} site.
</div>
{{< /site-region >}}

<div class="alert alert-warning">
Datadog .NET Profiler is currently in public beta. Datadog recommends evaluating the profiler in a non-sensitive environment before deploying in production.
</div>

## Requirements

**Supported operating systems:**

- Windows 10
- Windows Server starting from version 2012

**Supported .NET runtimes:**

64-bit applications running on:
- .NET Framework 4.6.1+
- .NET Core 2.1, 3.1
- .NET 5
- .NET 6

## Installation

<div class="alert alert-warning">

>   <strong>Note:</strong> Datadog's.NET Tracer and Profiler rely on the .NET CLR Profiling API. This API allows only one subscriber (for example, APM). To ensure maximum visibility, run only one APM solution in your application environment.

Reviewer: ```suggestion
  <strong>Note:</strong> Datadog's .NET Tracer and Profiler rely on the .NET CLR Profiling API. This API allows only one subscriber (for example, APM). To ensure maximum visibility, run only one APM solution in your application environment.
```

</div>

1. If you are already using Datadog, upgrade your agent to version [7.20.2][1]+ or [6.20.2][2]+.

> 2. The profiler and tracer are shipped together. Install or upgrade to version TODO+ if needed, using the [.NET Monitoring MSI installer][3]. Continuous Profiler only supports 64-bit Windows, with 32-bit support coming soon, so you'll want the file like `datadog-dotnet-apm-2.1.0-x64.msi`.

Reviewer: Commenting here to make sure we don't merge without fixing this TODO.

   ```
   {{< img src="tracing/setup/dotnet/RegistryEditorCore.png" alt="Using the Registry Editor to create environment variables for a Windows service" style="width:90%" >}}

   For .NET Framework:
   ```text
   COR_ENABLE_PROFILING=1
   COR_PROFILER={846F5F1C-F9AE-4B07-969E-05C26BC060D8}
   DD_ENV=prod
   DD_VERSION=1.2.3
   ```
   {{< img src="tracing/setup/dotnet/RegistryEditorFramework.png" alt="Using the Registry Editor to create environment variables for a Windows service" style="width:90%" >}}

   **With a PowerShell script:**

   For .NET Core and .NET 5+:
   ```powershell
   [string[]] $v = @(
       "CORECLR_ENABLE_PROFILING=1",
       "CORECLR_PROFILER={846F5F1C-F9AE-4B07-969E-05C26BC060D8}",
       "DD_ENV=prod",
       "DD_VERSION=1.2.3"
   )
   Set-ItemProperty HKLM:SYSTEM\CurrentControlSet\Services\<SERVICE NAME> -Name Environment -Value $v
   ```

   For .NET Framework:
   ```powershell
   [string[]] $v = @(
       "COR_ENABLE_PROFILING=1",
       "COR_PROFILER={846F5F1C-F9AE-4B07-969E-05C26BC060D8}",
       "DD_ENV=prod",
       "DD_VERSION=1.2.3"
   )
   Set-ItemProperty HKLM:SYSTEM\CurrentControlSet\Services\<SERVICE NAME> -Name Environment -Value $v
   ```

> 2. A minute or two after starting your application, your profiles will show up on the [Datadog APM > Profiler page][1].

Reviewer: ```suggestion
2. A minute or two after you start your application, your profiles appear on the [Datadog APM > Profiler page][1].
```

[1]: https://app.datadoghq.com/profiling
{{% /tab %}}

{{% tab "Standalone applications" %}}
1. To automatically profile a non-service application, such as console, ASP.NET (Core), Windows Forms, or WPF, some environment variables must be set before starting it.

>    If the environment variables are set for the current user, _all_ .NET applications will be profiled. Instead, we recommend setting them in a batch file that also starts the application and running that.

Reviewer: ```suggestion
   If the environment variables are set for the current user, _all_ .NET applications are profiled. Instead, set them in a batch file that also starts the application, and run the batch file.
```

   For .NET Core and .NET 5+:
   ```cmd
   SET CORECLR_ENABLE_PROFILING=1
   SET CORECLR_PROFILER={846F5F1C-F9AE-4B07-969E-05C26BC060D8}
   SET DD_ENV=prod
   SET DD_VERSION=1.2.3

   REM start the application here
   ```

   For .NET Framework:
   ```cmd
   SET COR_ENABLE_PROFILING=1
   SET COR_PROFILER={846F5F1C-F9AE-4B07-969E-05C26BC060D8}
   SET DD_ENV=prod
   SET DD_VERSION=1.2.3

   REM start the application here
   ```

> 2. A minute or two after starting your application, your profiles will show up on the [Datadog APM > Profiler page][1].

Reviewer: ```suggestion
2. A minute or two after you start your application, your profiles appear on the [Datadog APM > Profiler page][1].
```

| Environment variable                             | Type          | Description                                                                                      |
| ------------------------------------------------ | ------------- | ------------------------------------------------------------------------------------------------ |
| `DD_ENV`                   | String        | The [environment][4] name, for example, `production`. |
| `DD_SERVICE`               | String        | The [service][4] name, for example, `web-backend`. If this is not specified, the .NET Profiler tries to determine the service name automatically from the application name (process entry assembly or process name).    |
| `DD_VERSION`               | String        | The [version][4] of your service.  |
| `DD_TAGS`                  | String        | Tags to apply to an uploaded profile. Must be a list of `<key>:<value>` separated by commas such as: `layer:api,team:intake`.   |
| `DD_AGENT_HOST`            | String        | Sets the host where profiles are sent (the host running the Agent). Can be a hostname or an IP address. Ignored if `DD_TRACE_AGENT_URL` is set. Defaults to `localhost`.  |

> | `DD_TRACE_AGENT_PORT`      | String        | Sets the port where profiles are sent (the port where the Agent is listening for connections). Ignored if `DD_TRACE_AGENT_URL` is set. Defaults to`8126`..  |

Reviewer: ```suggestion
| `DD_TRACE_AGENT_PORT`      | String        | Sets the port where profiles are sent (the port where the Agent is listening for connections). Ignored if `DD_TRACE_AGENT_URL` is set. Defaults to`8126`.  |
```

    - link: 'tracing/profiler/profiler_troubleshooting'
      tag: 'Documentation'
      text: 'Fix problems you encounter while using the profiler'
---

{{< site-region region="us5" >}}
<div class="alert alert-warning">
  The Continuous Profiler is not available for the Datadog {{< region-param key="dd_site_name" >}} site.
</div>
{{< /site-region >}}

<div class="alert alert-warning">
Datadog .NET Profiler is currently in public beta. Datadog recommends evaluating the profiler in a non-sensitive environment before deploying in production.
</div>

<br>

## Requirements

**Supported operating systems:**

- Windows 10
- Windows Server starting from version 2012

**Supported .NET runtimes:**

64-bit applications running on:
- .NET Framework 4.6.1+
- .NET Core 2.1, 3.1
- .NET 5
- .NET 6

## Installation

<div class="alert alert-warning">
  <strong>Note:</strong> Datadog's .NET Tracer and Profiler rely on the .NET CLR Profiling API. This API allows only one subscriber (for example, APM). To ensure maximum visibility, run only one APM solution in your application environment.
</div>

1. If you are already using Datadog, upgrade your agent to version [7.20.2][1]+ or [6.20.2][2]+.

> 2. The profiler ships together with the tracer. Install or upgrade to the last beta version if needed, using the [.NET Monitoring MSI installer][3]. Continuous Profiler only supports 64-bit Windows, with 32-bit support coming soon, so you need the file like `datadog-dotnet-apm-2.1.0-x64-profiler-beta.msi`.

Reviewer: ```suggestion
2. The profiler ships together with the tracer. Install or upgrade to the last beta version if needed, using the [.NET Monitoring MSI installer][3]. Continuous Profiler only supports 64-bit Windows, so you need the file like `datadog-dotnet-apm-2.1.0-x64-profiler-beta.msi`.
```

   Run the .NET Monitoring MSI installer with administrator privileges.

> **Note**: Profiling requires the .msi version 2.1

Reviewer: ```suggestion
**Note**: Profiling requires the `.msi` version 2.1.
```

   DD_ENV=production
   DD_VERSION=1.2.3
   ```
   {{< img src="tracing/setup/dotnet/RegistryEditorFramework.png" alt="Using the Registry Editor to create environment variables for a Windows service" style="width:90%" >}}

   **With a PowerShell script:**

   For .NET Core and .NET 5+:
   ```powershell
   [string[]] $v = @(
       "CORECLR_ENABLE_PROFILING=1",
       "CORECLR_PROFILER={846F5F1C-F9AE-4B07-969E-05C26BC060D8}",
       "DD_PROFILING_ENABLED=1",
       "DD_ENV=production",
       "DD_VERSION=1.2.3"
   )
   Set-ItemProperty HKLM:SYSTEM\CurrentControlSet\Services\<SERVICE NAME> -Name Environment -Value $v
   ```

   For .NET Framework:
   ```powershell
   [string[]] $v = @(
       "COR_ENABLE_PROFILING=1",
       "COR_PROFILER={846F5F1C-F9AE-4B07-969E-05C26BC060D8}",
       "DD_PROFILING_ENABLED=1",
       "DD_ENV=production",
       "DD_VERSION=1.2.3"
   )
   Set-ItemProperty HKLM:SYSTEM\CurrentControlSet\Services\<SERVICE NAME> -Name Environment -Value $v
   ```

2. A minute or two after you start your application, your profiles appear on the [Datadog APM > Profiler page][1].

[1]: https://app.datadoghq.com/profiling
{{% /tab %}}

{{% tab "Standalone applications" %}}

> 1. Set needed environment variables to configure and enable Profiler for a non-service application, such as console, ASP.NET (Core), Windows Forms, or WPF. You can do this machine-wide, user-wide, or just for your service. We recommend doing this in a batch file that also starts the application, and running your application using the batch file.

Reviewer: ```suggestion
1. Set needed environment variables to configure and enable Profiler for a non-service application, such as console, ASP.NET (Core), Windows Forms, or WPF. You can do this machine-wide, user-wide, or just for your service. The recommended way to do this is in a batch file that also starts the application, and to run your application using the batch file.
```

  <strong>Note:</strong> Datadog's.NET Tracer and Profiler rely on the .NET CLR Profiling API. This API allows only one subscriber (for example, APM). To ensure maximum visibility, run only one APM solution in your application environment.
</div>

1. If you are already using Datadog, upgrade your agent to version [7.20.2][2]+ or [6.20.2][3]+.

> 2. Download the [.NET Monitoring MSI installer][1] version TODO+. Continuous Profiler only supports 64-bit, with 32-bit support coming soon.

Reviewer 2: We do this in other language pages so if, in the future, an apm user wants to enable profiling, if they're already running a new enough version, they can skip this step. There are pros and cons to this but if we want to get rid of it should do it for all languages.

3. Run the .NET Monitoring MSI installer with administrator privileges.

{{< tabs >}}

{{% tab "Internet Information Services (IIS)" %}}

> 1. Set environment variables:

Reviewer 2: We still want to suggest setting env, maybe service, and version for an optimal experience, so should include that here even though the installer sets the "enable the profiler" env vars.

         "DD_ENV=prod",
         "DD_SERVICE=my-web-app",
         "DD_VERSION=1.2.3"
     )
     Set-ItemProperty HKLM:SYSTEM\CurrentControlSet\Services\<SERVICE NAME> -Name Environment -Value $v
     ```
   - For .NET Framework:
     ```powershell
     [string[]] $v = @(
         "COR_ENABLE_PROFILING=1",
         "COR_PROFILER={846F5F1C-F9AE-4B07-969E-05C26BC060D8}",
         "DD_ENV=prod",
         "DD_SERVICE=my-web-app",
         "DD_VERSION=1.2.3"
     )
     Set-ItemProperty HKLM:SYSTEM\CurrentControlSet\Services\<SERVICE NAME> -Name Environment -Value $v
     ```

2. A minute or two after starting your application, your profiles will show up on the [Datadog APM > Profiler page][5].
{{% /tab %}}

{{% tab "Standalone applications" %}}
To automatically profile a non-service application, such as console, ASP.NET (Core), Windows Forms, or WPF, you'll need to set some environment variables before starting it.

If the environment variables are set for the current user, _all_ .NET applications will be profiled. Instead, we recomment setting them in a batch file that also starts the application and running that. For example:

 - For .NET Core and .NET 5+:
 ```

>  TODO show example batch file

Reviewer 2: Thanks!

 CORECLR_ENABLE_PROFILING=1
 CORECLR_PROFILER={846F5F1C-F9AE-4B07-969E-05C26BC060D8}
 DD_ENV=prod
 DD_SERVICE=my-web-app
 DD_VERSION=1.2.3
 ```

 - For .NET Framework:
 ```
 TODO show example batch file

 COR_ENABLE_PROFILING=1
 COR_PROFILER={846F5F1C-F9AE-4B07-969E-05C26BC060D8}
 DD_ENV=prod
 DD_SERVICE=my-web-app
 DD_VERSION=1.2.3
 ```

2. A minute or two after starting your application, your profiles will show up on the [Datadog APM > Profiler page][5].
{{% /tab %}}
{{< /tabs >}}

## Configuration

You can configure the profiler using the following environment variables. Restart the application after any of these settings is changed.

> | Environment variable                             | Type          | Description                                                                                      |

Reviewer 2: Gotcha. Which tracer doc are you comparing with? I think there are a couple.
